File: minSwapsSort.py
def lilysHomework(arr):
    # function to count swaps to bring sorted array (s)
    # to reference array (r)
    def countSwaps(s,ref):
        count = 0
        i=0
        while i < len(arr):
        # if reference and sorted to the same value
            if s[i]['val'] != ref[i]['val']:
        # swap them and increment count
        # don't increment i to check again
                s[s[i]['idx']],s[i] = s[i],s[s[i]['idx']]
                count += 1
            else:
                i += 1

        return count

    # create object of value and index
    vec = [{"val": arr[i], "idx": i} for i in range(len(arr))]
    # assending sorted array
    a = sorted(vec, key=lambda x: x['val'])
    # dessending sorted array
    d = sorted(vec, key=lambda x: x['val'], reverse=True)

    return min(countSwaps(a,vec), countSwaps(d,vec))

lilysHomework([5,3,1,4,2])    

File: javascript/minSwapsSort.js
function lilysHomework(arr) {
	// function to count swaps to bring sorted array (s)
	// to reference array (r)
	function countSwaps(s) {
		let count = 0;
		let i = 0;
		while (i < arr.length) {
			if (s[i].val != arr[i]) {
				// swap them and increment count
				// don't increment i to check again
				[s[s[i].idx], s[i]] = [s[i], s[s[i].idx]];
				count += 1;
			} else i += 1;
		}

		return count;
	}

	// create object of value and index
	const vec = arr.map((val, i) => {
		return { val: val, idx: i };
	});
	// assending sorted array
	a = [...vec].sort((a, b) => a.val - b.val);
	// dessending sorted array
	d = [...vec].sort((a,b) => b.val - a.val)

	return Math.min(countSwaps(a), countSwaps(d));
}
console.log(lilysHomework([5, 3, 1, 4, 2]));
